### PR TITLE
Migrate s3io to jobrunner

### PR DESCRIPTION
**Pull request by ericmjonas**
Right now we save func and data to disk and THEN invoke jobrunner, with the idea of jobrunner being as simple as possible. As a result, the size of func + data + runtime < worker /tmp space. It's not unreasonable to expect data to sometimes be large. We then load func and data into memory anyway. This PR shifts reading those items from S3 into `jobrunner.py`. 

Advantages:
- Now func and data can be as large as memory! 
- Wrenhandler becomes simpler

Disadvantages:
- Slightly harder to get logging information out of jobrunner for when access fails. 
- Harder to get s3 download metrics out as well

I wanted to solicit feedback on this early.  I think this is a much better approach and IIRC @Vaishaal was explicitly asking for it.

### CODE REVIEW

**Comment by shivaram** [comment]
@ericmjonas Do we want this for 0.3 or can we take some more time to review this

**Comment by ericmjonas** [comment]
We can totally take more time; i don't think it's necessary for 0.3

File: pywren/wrenhandler.py [hunk 1]
```diff
@@ -294,11 +294,21 @@ def generic_handler(event, context_dict):
         CONDA_PYTHON_PATH = "/tmp/condaruntime/bin"
         CONDA_PYTHON_RUNTIME = os.path.join(CONDA_PYTHON_PATH, "python")
 
-        cmdstr = "{} {} {} {} {}".format(CONDA_PYTHON_RUNTIME,
-                                         jobrunner_path,
-                                         func_filename,
-                                         data_filename,
-                                         output_filename)
+        # pass a full json blob
+        jobrunner_config_filename = "/tmp/jobrunner.config.json"
```
Comment (shivaram) [inline pywren/wrenhandler.py]: can we make this a constant at the top of the file ? Also generating a unique name might be good ?

File: pywren/wrenhandler.py [hunk 2]
```diff
@@ -244,33 +244,33 @@ def generic_handler(event, context_dict):
         logger.info("data data download complete, took {:3.2f} sec".format(data_download_time))
         response_status['data_download_time'] = data_download_time
 
-        # now split
-        d = json.load(open(func_filename, 'r'))
+        # d = json.load(open(func_filename, 'r'))
```
Comment (shivaram) [inline pywren/wrenhandler.py]: So I'm not sure I fully follow why we still have the data_key download and func_key download happening in wrenhandler -- isn't the func object the thing we want to move to memory as it gets to 25MB or so from notebooks ?

File: pywren/jobrunner.py [hunk 3]
```diff
@@ -1,33 +1,88 @@
...
+    jobrunner_config_filename = sys.argv[1]
+    out_filename = sys.argv[2]
+
+
     # initial output file in case job fails
     pickle.dump({'result' : None,
                  'success' : False},
                 open(out_filename, 'wb'), -1)
+    jobrunner_config = json.load(open(jobrunner_config_filename,
+                                      'r'))
```
Comment (shivaram) [inline pywren/jobrunner.py]: really minor nit - i think the `r` can go in the previous line

File: pywren/jobrunner.py [hunk 4]
```diff
@@ -1,33 +1,88 @@
...
+
+
+    # FIXME someday switch to storage handler
+    # download the func data into memory
+    s3_client = boto3.client("s3")
+    func_bucket, func_key = jobrunner_config['func_bucket'], jobrunner_config['func_key']
+    data_bucket, data_key = jobrunner_config['data_bucket'], jobrunner_config['data_key']
+
+    data_byte_range = jobrunner_config['data_byte_range']
+    func_obj_stream = s3_client.get_object(Bucket=func_bucket, Key=func_key)
```
Comment (shivaram) [inline pywren/jobrunner.py]: this is already supported by the storage API right ? is there any reason we can't directly use it ?

File: pywren/executor.py [hunk 5]
```diff
@@ -194,14 +194,14 @@ def map(self, func, iterdata, extra_env=None, extra_meta=None,
...
-        func_str_encoded = wrenutil.bytes_to_b64str(func_str)
+        func_str_encoded = func_str
         #debug_foo = {'func' : func_str_encoded,
         #             'module_data' : module_data}
 
         #pickle.dump(debug_foo, open("/tmp/py35.debug.pickle", 'wb'))
         ### Create func and upload
-        func_module_str = json.dumps({'func' : func_str_encoded,
-                                      'module_data' : module_data})
+        func_module_str = pickle.dumps({'func' : func_str_encoded,
```
Comment (shivaram) [inline pywren/executor.py]: is there a reason for switching from json to pickle here ? its a bit weird as we already have the func_str pickled if i'm reading the code right

File: pywren/jobrunner.py [hunk 6]
```diff
@@ -1,33 +1,88 @@
...
+    loaded_func = pickle.loads(loaded_func_all['func'])
+
+    extra_get_args = {}
+    if data_byte_range is not None:
+        range_str = 'bytes={}-{}'.format(*data_byte_range)
+        extra_get_args['Range'] = range_str
+    data_obj_stream = s3_client.get_object(Bucket=data_bucket,
+                                           Key=data_key, **extra_get_args)
+    # FIXME make this streaming
+    loaded_data = pickle.loads(data_obj_stream['Body'].read())
```
Comment (shivaram) [inline pywren/jobrunner.py]: why is this not streaming right now ? you mean you want to stream the data into the function ?

File: pywren/wrenhandler.py [hunk 7]
```diff
@@ -244,33 +244,33 @@ def generic_handler(event, context_dict):
...
-            full_filename = os.path.join(to_make, os.path.basename(m_filename))
-            #print "creating", full_filename
-            fid = open(full_filename, 'wb')
-            fid.write(b64str_to_bytes(m_data))
-            fid.close()
-        logger.info("Finished writing {} module files".format(len(d['module_data'])))
-        logger.debug(subprocess.check_output("find {}".format(PYTHON_MODULE_PATH), shell=True))
-        logger.debug(subprocess.check_output("find {}".format(os.getcwd()), shell=True))
+
+        # # get modules and save
```
Comment (shivaram) [inline pywren/wrenhandler.py]: again lets delete instead of commenting :)

File: pywren/executor.py [hunk 8]
```diff
@@ -194,14 +194,14 @@ def map(self, func, iterdata, extra_env=None, extra_meta=None,
                         mod_paths.remove(mod_path)
 
         module_data = create_mod_data(mod_paths)
-        func_str_encoded = wrenutil.bytes_to_b64str(func_str)
+        func_str_encoded = func_str
```
[the remark this answers is not in the document]
Comment (ericmjonas) [reply pywren/executor.py]: yes!

File: pywren/executor.py (hunk 5 above)
Comment (ericmjonas) [reply pywren/executor.py]: we first want to unpickle the modules and get them set up in the `PYTHONPATH` before we unpickle the function, otherwise we will get an error unpickling. Thus the two-stage process.

Comment (shivaram) [reply pywren/executor.py]: The two stage process is fine -- i was just curious why we changed the second layer or serialization from json to pickle. I might be missing something basic ?

File: pywren/jobrunner.py (hunk 6 above)
Comment (ericmjonas) [reply pywren/jobrunner.py]: We're calling `.read()` which reads the entire object into memory as a string, and then `loads` which unpickles the string. The "right" thing to do would be pass the `data_object_stream['Body']` to `pickle.load` which wants a file-like object. The problem is that the streaming file-like object boto provides doesn't actually provide _enough_ of the file object API for pickle to use it. We could have a wrapper (like we do for the call to `tarfile`) but even that one is insufficient. 

So for N bytes of input data this means that we briefly have 2N. That said, the temporary string created is easy for GC to get rid of, and this is currently an order of magnitude better than what we had been storing in /tmp/

File: pywren/wrenhandler.py (hunk 2 above)
Comment (ericmjonas) [reply pywren/wrenhandler.py]: Sorry, again, this is in-progress, there's some detritus that needs to be cleaned out. I should have been more clear, i'm sorry!

File: pywren/executor.py (hunk 5 above)
Comment (ericmjonas) [reply pywren/executor.py]: We only were using json because we needed wrenhandler, which runs in the lambda-supplied python (2.7 until recently) to write the `foo.py` module files, and  pickle isn't stable cross-python versions. So we needed something independent there, and then internally to that json object we were storing the base64-encoded pickled function -- because pickling results in a binary stream and we then need to base64 it for json. 

In short, pickle is faster to deserialize, more space-efficient, prevents us from having to b64-encode the internal pickled object, and appropriate because the remote `jobrunner` is running the same version of python as the user.

**Comment by apengwin** [comment]
Why don't we write the `output.pickle` directly from s3 from jobrunner, instead of writing it to /tmp?

**Comment by ericmjonas** [comment]
@apengwin That's a REALLY GOOD QUESTION we should do that. Thank you!

File: pywren/wrenhandler.py [hunk 9]
```diff
@@ -211,67 +204,13 @@ def generic_handler(event, context_dict):
...
-
-        data_download_time = time.time() - start_time
-        logger.info("data data download complete, took {:3.2f} sec".format(data_download_time))
-        response_status['data_download_time'] = data_download_time
-
-        # now split
-        d = json.load(open(func_filename, 'r'))
+        response_status['free_disk_bytes'] = free_disk_bytes
+        # clean up for modules
         shutil.rmtree(PYTHON_MODULE_PATH, True) # delete old modules
```
Comment (shivaram) [inline pywren/wrenhandler.py]: should we also move this inside the jobrunner ? I'm not sure it matters but it'll make the behavior same as before (i.e. download func and if that succeeds then do rmtree)

Comment (ericmjonas) [reply pywren/wrenhandler.py]: yeah good catch will do
